File: minicore/src/block_allocator.cpp
#include "minicore/block_allocator.hpp"

#include <algorithm>

namespace minicore {

BlockAllocator::BlockAllocator(int num_blocks, int block_size)
    : num_blocks_(num_blocks), block_size_(block_size) {
  if (num_blocks < 1 || block_size < 1) {
    throw std::invalid_argument("num_blocks and block_size must be positive, got " +
                                std::to_string(num_blocks) + ", " + std::to_string(block_size));
  }
  free_.resize(static_cast<std::size_t>(num_blocks));
  // Reversed, so that back() (popped first) is block 0: same order as the
  // reference implementation's `list(range(num_blocks - 1, -1, -1))`.
  for (int i = 0; i < num_blocks; ++i) free_[static_cast<std::size_t>(i)] = num_blocks - 1 - i;
  ref_.assign(static_cast<std::size_t>(num_blocks), 0);
  drops_.assign(static_cast<std::size_t>(num_blocks), 0);
}

void BlockAllocator::allocate_into(int n, std::vector<BlockId>& out) {
  if (n < 0) throw std::invalid_argument("cannot allocate " + std::to_string(n) + " blocks");
  if (n > static_cast<int>(free_.size())) {
    throw OutOfBlocks("requested " + std::to_string(n) + " blocks, " + std::to_string(free_.size()) +
                      " free");
  }
  out.reserve(out.size() + static_cast<std::size_t>(n));
  for (int i = 0; i < n; ++i) {
    BlockId b = free_.back();
    free_.pop_back();
    ref_[static_cast<std::size_t>(b)] = 1;
    out.push_back(b);
  }
}

std::vector<BlockId> BlockAllocator::allocate(int n) {
  std::vector<BlockId> out;
  allocate_into(n, out);
  return out;
}

void BlockAllocator::incref(const std::vector<BlockId>& blocks) {
  for (BlockId b : blocks) {
    check_id(b);
    if (ref_[static_cast<std::size_t>(b)] == 0) {
      throw std::invalid_argument("incref of free block " + std::to_string(b));
    }
  }
  for (BlockId b : blocks) ref_[static_cast<std::size_t>(b)] += 1;
}

void BlockAllocator::free(const std::vector<BlockId>& blocks) {
  for (BlockId b : blocks) check_id(b);
  // Count releases per block, then check each against its reference count.
  // Touched entries are reset below whatever happens, so `drops_` is all zeros
  // on entry and on exit, including when this throws.
  for (BlockId b : blocks) drops_[static_cast<std::size_t>(b)] += 1;
  BlockId offender = -1;
  std::int32_t releases = 0;
  for (BlockId b : blocks) {
    std::int32_t k = drops_[static_cast<std::size_t>(b)];
    if (k > 0 && ref_[static_cast<std::size_t>(b)] < k) {
      offender = b;
      releases = k;
      break;
    }
  }
  for (BlockId b : blocks) drops_[static_cast<std::size_t>(b)] = 0;
  if (offender >= 0) {
    throw std::invalid_argument("free of block " + std::to_string(offender) + ": " +
                                std::to_string(releases) + " releases but reference count " +
                                std::to_string(ref_[static_cast<std::size_t>(offender)]));
  }
  for (BlockId b : blocks) {
    if (--ref_[static_cast<std::size_t>(b)] == 0) free_.push_back(b);
  }
}

void BlockAllocator::check_invariants() const {
  std::vector<char> seen(static_cast<std::size_t>(num_blocks_), 0);
  for (BlockId b : free_) {
    if (seen[static_cast<std::size_t>(b)]) {
      throw InvariantError("duplicate block in free stack");
    }
    seen[static_cast<std::size_t>(b)] = 1;
  }
  for (int b = 0; b < num_blocks_; ++b) {
    bool is_free = seen[static_cast<std::size_t>(b)] != 0;
    if (is_free != (ref_[static_cast<std::size_t>(b)] == 0)) {
      throw InvariantError("free stack disagrees with reference counts");
    }
    if (ref_[static_cast<std::size_t>(b)] < 0) throw InvariantError("negative reference count");
  }
}

}  // namespace minicore

File: miniserve/cache/block_allocator.py
"""Fixed-size KV block allocator with per-block reference counts.

The KV pool is divided into ``num_blocks`` blocks of ``block_size`` tokens.
This class only tracks which blocks are free and how many holders each used
block has; it owns no tensors.

Allocation order is part of the contract, so that other implementations can
be checked block id for block id: a fresh allocator hands out 0, 1, 2, ...;
freed blocks go on top of a stack and are reused first (LIFO).
"""

from __future__ import annotations

from collections.abc import Iterable


class OutOfBlocks(RuntimeError):
    pass


BACKEND = "python"


class BlockAllocator:
    def __init__(self, num_blocks: int, block_size: int):
        if num_blocks < 1 or block_size < 1:
            raise ValueError(f"num_blocks and block_size must be positive, got {num_blocks}, {block_size}")
        self.num_blocks = num_blocks
        self.block_size = block_size
        self._free = list(range(num_blocks - 1, -1, -1))  # stack: pop() yields 0 first
        self._ref = [0] * num_blocks  # 0 means free

    @property
    def num_free(self) -> int:
        return len(self._free)

    @property
    def num_used(self) -> int:
        return self.num_blocks - len(self._free)

    def can_allocate(self, n: int) -> bool:
        return 0 <= n <= len(self._free)

    def refcount(self, block: int) -> int:
        self._check_id(block)
        return self._ref[block]

    def allocate(self, n: int) -> list[int]:
        """Take ``n`` free blocks (reference count 1 each). All or nothing."""
        if n < 0:
            raise ValueError(f"cannot allocate {n} blocks")
        if n > len(self._free):
            raise OutOfBlocks(f"requested {n} blocks, {len(self._free)} free")
        out = [self._free.pop() for _ in range(n)]
        for b in out:
            self._ref[b] = 1
        return out

    def incref(self, blocks: Iterable[int]) -> None:
        """Add a holder to each block. Blocks must be allocated."""
        blocks = list(blocks)
        for b in blocks:
            self._check_id(b)
            if self._ref[b] == 0:
                raise ValueError(f"incref of free block {b}")
        for b in blocks:
            self._ref[b] += 1

    def free(self, blocks: Iterable[int]) -> None:
        """Drop one holder from each block; blocks with no holders left return to the pool.

        The whole call is validated before anything changes. A block may appear
        more than once only if it has at least that many holders.
        """
        blocks = list(blocks)
        drops: dict[int, int] = {}
        for b in blocks:
            self._check_id(b)
            drops[b] = drops.get(b, 0) + 1
        for b, k in drops.items():
            if self._ref[b] < k:
                raise ValueError(f"free of block {b}: {k} releases but reference count {self._ref[b]}")
        for b in blocks:
            self._ref[b] -= 1
            if self._ref[b] == 0:
                self._free.append(b)

    def new_table(self, prefix_blocks: Iterable[int] = ()) -> "BlockTable":
        """A block table over this allocator, optionally starting with a shared prefix.

        Callers use this instead of naming a table class, so that a table always
        matches the backend of the allocator it is built on.
        """
        from miniserve.cache.block_table import BlockTable

        return BlockTable(self, list(prefix_blocks))

    def check_invariants(self) -> None:
        """Free stack and reference counts describe the same set; raises AssertionError otherwise."""
        free = set(self._free)
        assert len(free) == len(self._free), "duplicate block in free stack"
        assert free == {b for b, r in enumerate(self._ref) if r == 0}, "free stack disagrees with reference counts"
        assert all(r >= 0 for r in self._ref), "negative reference count"

    def _check_id(self, block: int) -> None:
        if not 0 <= block < self.num_blocks:
            raise IndexError(f"block id {block} out of range [0, {self.num_blocks})")
